Mc rss update

maxachis opened the pull request:
## What I did

- Renamed from rss-etl to rss_etl (easier to parse)
- Revised rss_etl to be more class-based and reduce redundant code
- Revised rss_etl to enable inserting only a certain number of new entries at a time, rather than attempting to insert all at once
- Revised rss_etl by adding methods for inserting to Bills table (previously was only Revisions)
- Added initial test methods for rss_etl
- Add Supabase directory with necessary supabase files for setting up local dev environment.

## Why I did it

Done in accordance with https://trello.com/c/7viNV7Kq/48-update-rssetlpy-to-add-data-to-bills-table-in-supabase-as-well, while also setting up some of the initial foundation for https://trello.com/c/kqg0J6Ei/38-supabase-dev-environment-for-back-end

## Issue #s Addressed

https://trello.com/c/7viNV7Kq/48-update-rssetlpy-to-add-data-to-bills-table-in-supabase-as-well

- closes #

## How to test it

- Setup local Supabase dev environment, setup local .env file to point to local supabase dev and comment/uncomment code in rss_etl controlling whether api url and key are being pulled from Github actions or local environment config.

maxachis committed:
Add Bills table insertion, testing for rss_etl:
- Renamed from rss-etl to rss_etl (easier to parse)
- Revised to be more class-based and reduce redundant code
- Added methods for inserting to Bills table (previously was only Revisions)
- Added initial test method (likely to need to be changed).

maxachis committed:
Modified test methods, fixed bug involving the same object being overwritten in memory.

maxachis committed:
Updated rss_etl to account for new DB design and to ensure bill internal id's are obtained before adding revision to revision table.

maxachis committed:
Add supabase dev environment support, update db insertion functionality

Added Supabase directory with necessary supabase files for setting up local dev environment.

Modified sb_api_url variable to also be drawn from Github Actions Secret Management, more for consistency's sake than any security concern.

Update "create_and_attempt_to_insert" functions along with "extract_metadata_from_rss_feed" such that they are set to insert only a set number of new entries (a new entry being defined as a new revision to a bill or a new bill-revision combination), and will stop after that set number of new entries is inserted.

mikeandike523 commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+        return api_response.data[0]["bill_internal_id"]
+
+    def create_revisions_output_record(self, bill, bill_identifier: BillIdentifier):
+        revisions_output_record = OutputRecord(self.supa_con, "Revisions", REVISIONS_OUTPUT_DICT)
```
Nit:

instead of having multiple subsequent assignments, you can use `dict.update`

E.g.
```python
        revisions_output_record.output_dict["bill_internal_id"] = self.get_bill_internal_id(bill_identifier)
        revisions_output_record.output_dict["printer_no"] = bill_identifier.printer_number
        revisions_output_record.output_dict["full_text_link"] = bill["link"]
        revisions_output_record.output_dict["publication_date"] = bill["published"]
        revisions_output_record.output_dict["revision_guid"] = bill["guid"]
        revisions_output_record.output_dict["description"] = bill["description"]
```

becomes

```python
revisions_output_record.output_dict.update({
    "bill_internal_id": self.get_bill_internal_id(bill_identifier),
    "printer_no": bill_identifier.printer_number,
    "full_text_link": bill["link"]

     #etc...

})
```


})
```

maxachis replied:
Addressing this in my upcoming commit!

maxachis commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+# sb_api_url = os.getenv("SUPABASE_API_URL")  # local secret management with .env file
+# sb_api_key = os.getenv("SUPABASE_API_KEY")  # local secret management with .env file
+sb_api_url = os.environ["SUPABASE_API_URL"]  # github actions secret management
+sb_api_key = os.environ["SUPABASE_API_KEY"]  # github actions secret management
```
Don't need both versions of this code --if you have one, you don't need the other.

maxachis commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+
+class LegislativeBody(Enum):
+    SENATE = "Senate"
+    HOUSE = "House"
```
In the future, move this to a separate constants file

maxachis replied:
Ehhh, on second thought I'll leave it as is right now, with intent to reevaluate later if it proves necessary. The script is 250 lines of code, and we're already doing a lot in this PR. In a future PR, maybe.

maxachis commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+}
+
+
+class BillIdentifier:
```
Generally, have separate files for the classes, separate from the main execution.

maxachis replied:
Again, I'll notch this for a separate PR -- there's already a lot going on here.

mikeandike523 commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+
+    @staticmethod
+    def determine_chamber(chamber_letter):
+        if chamber_letter == "H":
```
you forgot to use the legislative body enum `LegislativeBody`. Why define the enum and not use it everywhere you can. In python, "enums" support equality checking so you can use them pretty freely.

maxachis replied:
Noted! Accounted for in upcoming commit!

maxachis commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+        self.supa_con.table(out_rec.table_name).insert(out_rec.output_dict).execute()
+
+
+def extract_from_rss_feed(supabase_connection, leg_bod, rss_feed):
```
Generally, this line and everything else could just be kept in this file, and everything else could be helper files.

maxachis commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+
+    def __init__(self, guid):
+        self.guid = guid
+        self.bill_guid = re.search(r"(\d{4}\d[H|S]B\d+)P\d+", guid).group(1)
```
For this regex; just add comments describing what it's doing and give an example

mikeandike523 commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+
+    def has_existing_supabase_records(self, matching_columns):
+        # TODO: This might suffice for now, but selecting every row in the table is not a good long-term solution
+        statement = self.supa_con.table(self.table_name).select("*", count="exact")
```
talked to chat GPT 4 about how do the count query efficiently. Easy to change

Simply use:
```python
statement=self.sup_con.table(self.table_name).select("count(*)")
for column in matching_columns:
    statement = statement.eq(column, self.output_dict[column])
    response = statement.execute()
    if response.status_code = 200:
        count = response.data[0]['count']
    else:
        raise Exception(f"""
        Failed to query supabase:
        status_code: {response.status_code}
        error_message: {response.error_message}
        """)
```

jmeaster30 replied:
How is that code different than the code that is already there? It looks like it executes the query multiple times so you make several db calls instead of just one

maxachis replied:
Agreed with John -- this result doesn't look too different, and querying for every column individually would just add overhead. 
In retrospect, since it's just executing a where clause, it should be fine.

I do like the error handling, however, and will include that in my next commit.

maxachis replied:
Actually, after trying to work it out, that particular error handling doesn't work. Honestly, when I had the API break while testing it, I found the errors sufficiently descriptive, so I'll leave that be for now.

mikeandike523 replied:
Sorry i made a mistake with the indentation in my suggestion. The `.execute()` should only run after the loop is done
statement=self.sup_con.table(self.table_name).select("count(*)")
```python
for column in matching_columns:
    statement = statement.eq(column, self.output_dict[column])
response = statement.execute()
if response.status_code == 200:
    count = response.data[0]['count']
else:
    raise Exception(f"""
    Failed to query supabase:
    status_code: {response.status_code}
    error_message: {response.error_message}
    """)
```

maxachis replied:
That still didn't work. statement.execute() doesn't return something with a "status_code" attribute. I think you're thinking of http status code, but this functions differently.

mikeandike523 commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+
+    def __init__(self, supa_con, leg_body, rss_feed):
+        self.supa_con = supa_con
+        self.leg_body = leg_body
```
Nit:

Why not just call it "legislative_body" instead of "leg_body", it's much clearer and the line won't be too long

maxachis replied:
Good call -- added in next commit.

mikeandike523 commented on src/rss_etl.py:
A few large-scale improvements that are worthwhile:

1. Handling errors, especially for supabase queries.
2. Adding type annotations
3. More docstring, even for some class properties if relevant
4. More comments for code that is difficult to understand

maxachis replied:
Addressing this in next commit!

jmeaster30 commented on src/rss_etl.py:
```diff
@@ -0,0 +1,175 @@
+
+    def __init__(self, guid):
+        self.guid = guid
+        self.bill_guid = re.search(r"(\d{4}\d[H|S]B\d+)P\d+", guid).group(1)
```
This regex threw an exception for me while trying this out. The above re.search call returns "None" if the regex fails and the rss feed gave what I am guessing is a revision instead of a bill so it had an 'R' where a 'B' was expected. Is the BillIdentifier class meant to handle both bills and revisions?

jmeaster30 replied:
For my local testing I changed the regexes to something like `(\d{4}\d[H|S](?:B|R)\d+)P\d+` just to not mess with the groups. Also the `[H|S]` matches an 'H', an 'S', or a '|'. With regexes everything in square brackets is treated as a literal character (expect '^' and '-' in certain cases)

maxachis replied:
> This regex threw an exception for me while trying this out. The above re.search call returns "None" if the regex fails and the rss feed gave what I am guessing is a revision instead of a bill so it had an 'R' where a 'B' was expected. Is the BillIdentifier class meant to handle both bills and revisions?

All RSS entries should be bill revisions, rather than bills itself. This is in part because bills have a one-to-many relationship with revisions -- one bill can have multiple revisions, but the opposite is not true. A bill must have at least one revision.

The R in that case did not refer to a Revision, but a _Resolution_. This is unclear in my code, and I am updating it accordingly as part of my attempts to explain the regex better overall.

> For my local testing I changed the regexes to something like `(\d{4}\d[H|S](?:B|R)\d+)P\d+` just to not mess with the groups. Also the `[H|S]` matches an 'H', an 'S', or a '|'. With regexes everything in square brackets is treated as a literal character (expect '^' and '-' in certain cases)

Excellent catch! I've corrected this in my own code as well.

maxachis committed:
Removed unused methods
Added descriptive commenting
Added type-hinting
Corrected regex error in BillIdentifier Class
Added error-catching in BillIdentifier Class
Renamed some variables and class names for clarity

There was a debate over whether to split rss_etl into separate scripts for better readability. After much thought and prayer I concluded that, for the moment, I'll keep everything related to the rss_etl in one script, with the option to migrate and separate them later. The code is 250 lines, and everything is closely related.

In terms of error handling, I added additional error catching only for the regex in BillIdentifier, mostly because the basic regex error is confusing and nondescriptive. With everything else, I'm content to leave things as is, with intent to reevaluate later if we get a lot of noisy breaking, or people propose some useful error catching.

mikeandike523 commented on src/test_rss_etl.py:
```diff
@@ -11,6 +11,10 @@ class TestExtractor(TestCase):
 
+    def print_call_output_dict(self, call_args_list):
+        for idx, call in enumerate(call_args_list):
+            print(f"{idx}: Insert to {call[0][0].table_name}: {call[0][0].output_dict}")
```
Not sure why its `call[0][0]` and not `call[idx][0]`

mikeandike523 commented on src/rss_etl.py:
```diff
@@ -64,39 +108,49 @@ def extract_metadata_from_rss_feed(self, max_attempts=3):
-    def create_and_attempt_to_insert_revision(self, bill, bill_no, printer_no):
-        revisions_output_record = self.create_revisions_output_record(bill, bill_no, printer_no)
-        if not revisions_output_record.has_existing_supabase_records(["bill_id", "printer_no", "legislative_body"]):
+    def create_and_attempt_to_insert_revision(self, bill, bill_identifier: BillIdentifier):
```
NIT

Using "upsert" may be clearer and more concise. I.e. `def upsert_revision`. You can apply similar renames to other functions if desired

maxachis replied:
"upsert" would not be quite descriptive of what's going on here -- there's no update occurring if a row already exists.

mikeandike523 commented on src/rss_etl.py:
```diff
@@ -64,39 +108,49 @@ def extract_metadata_from_rss_feed(self, max_attempts=3):
         revisions_output_record = OutputRecord(self.supa_con, "Revisions", REVISIONS_OUTPUT_DICT)
-        revisions_output_record.output_dict["bill_id"] = bill_no
-        revisions_output_record.output_dict["printer_no"] = printer_no
+        revisions_output_record.output_dict["bill_internal_id"] = self.get_bill_internal_id(bill_identifier)
```
We need to finalize our choices for what the revisions/bills table primary key will be. Are we using the GUID provided from the RSS feed, or a randomly generated one? What is the relationship between `bill_id` and `bill_internal_id`? If you want to make it clear, though  this will require discussion between mutliple people, I recommend calling it `bill_rss_guid` or `bill_gov_guid` if using the rss one, or if generating it, call it `bill_generated_id`, or `bill_postgres_id` or `bill_primary_key`, etc.

In my personal opinion, we should use the GUID provided by the government/found in the RSS feed. Because if we try to make our own GUID on top of that to resolve some sort of database integrity error, we are essentially making up information that isn't there. In the case of conflict, we should just log it as an error and move on. We cannot solve the governments organizational issues

maxachis replied:
The code you're referencing is outdated -- in the new version, I've removed the "bill_id" and made the column nomenclature more clear. 

I also agree that we should use the GUID as the primary key. The issue is whether we should have a numeric internal ID for linking, or if the GUID is sufficient. 

But more prominently, to what degree do we want to block moving forward on this because of that? If we decide we want to get rid of the bill_internal_id, that would be a relatively trivial change in the database and these scripts -- we'd simply substitute them for the guid.

mikeandike523 commented on src/rss_etl.py:
```diff
@@ -55,121 +56,197 @@ class BillIdentifier:
+        :return: True if a new record was inserted, False if not
+        """
+        revisions_output_record = self.create_revisions_output_record(revision_rss_feed_entry, bill_identifier)
         if not revisions_output_record.has_existing_supabase_records(["revision_guid"]):
```
NIT

I don't think this naming convention is good. How can a record have an existing record?

maxachis replied:
Agreed. Revised.

mikeandike523 commented on src/rss_etl.py:
```diff
@@ -55,121 +56,197 @@ class BillIdentifier:
-    def get_bill_internal_id(self, bill_identifier: BillIdentifier):
-        statement = self.supa_con.table("Bills").select("bill_internal_id").eq("legislative_id",
-                                                                               bill_identifier.bill_guid)
+    def get_bill_internal_id(self, bill_identifier: BillIdentifier) -> str:
```
once again, we need to decide as a team if `legistlative_id` will be sufficient and `bill_id` / `bill_internal_id` is not needed

maxachis committed:
Rename InsertRecord method to "exists_in_supabase" from "has_existing_supabase_records"

maxachis commented:
I'm going to go ahead and merge this now. My rationale is that I implemented most of the changes people suggested and I didn’t want to busy up the next meeting by reviewing the same code. If people find there are issues with the changes merged, we can always make changes to it later on and, in a pinch, revert the commit. 

I’m also revising/formalizing pull request standards so that we can get PRs moving faster and avoiding stalls.
